scvi/hub/hub_model.py

```diff
@@ -0,0 +1,284 @@
+        api.upload_file(
+            path_or_fileobj=json.dumps(asdict(self.metadata), indent=4).encode(),
+            path_in_repo=_SCVI_HUB.METADATA_FILE_NAME,
+            repo_id=repo_name,
+            token=repo_token,
+        )
+
+    @classmethod
+    def pull_from_huggingface_hub(cls, repo_name: str):
+        """
+        Download the given model repo from huggingface.
+
+        The model, its card, data, metadata are downloaded to a cached location on disk
+        selected by huggingface and an instance of this class is created with that info
+        and returned.
+
+        Parameters
+        ----------
+        repo_name
+            ID of the huggingface repo where this model needs to be uploaded
+        """
+        cache_dir = snapshot_download(
+            repo_id=repo_name,
+            allow_patterns=["model.pt", "adata.h5ad", _SCVI_HUB.METADATA_FILE_NAME],
+        )
+        model_card = ModelCard.load(repo_name)
+        return cls(cache_dir, model_card=model_card)
+
+    def __repr__(self):
+        def eval_obj(obj):
+            return "No" if obj is None else "Yes"
+
+        print(
+            "HubModel with:\n"
+            f"local_dir: {self._local_dir}\n"
+            f"model loaded? {eval_obj(self._model)}\n"
+            f"adata loaded? {eval_obj(self._adata)}\n"
+            f"large_training_adata loaded? {eval_obj(self._large_training_adata)}\n"
+            f"metadata:\n{self.metadata}\n"
+            f"model_card:"
+        )
+        rich.print(Markdown(self.model_card.content))
+        return ""
+
+    @property
+    def metadata(self) -> HubMetadata:
+        """The metadata for this model."""
+        return self._metadata
+
+    @property
+    def model_card(self) -> ModelCard:
+        """The model card for this model."""
+        return self._model_card
+
+    @property
+    def model(self) -> Type[BaseModelClass]:
+        """
+        Returns the model object for this hub model.
+
+        If the model has not been loaded yet, this will call :meth:`~scvi.hub.HubModel.load_model`.
+        Otherwise, it will simply return the loaded model.
+        """
+        if self._model is None:
+            self.load_model()
+        return self._model
+
+    @property
+    def adata(self) -> Optional[AnnData]:
+        """
+        Returns the data for this model.
+
+        If the data has not been loaded yet, this will call :meth:`~scvi.hub.HubModel.read_adata`.
+        Otherwise, it will simply return the loaded data.
+        """
+        if self._adata is None:
+            self.read_adata()
+        return self._adata
+
+    @property
+    def large_training_adata(self) -> Optional[AnnData]:
+        """
+        Returns the training data for this model, which might be too large to reside within the hub model.
+
+        If the data has not been loaded yet, this will call :meth:`~scvi.hub.HubModel.read_large_training_adata`.
+        Otherwise, it will simply return the loaded data.
```

Comment:
```suggestion
        If the data has not been loaded yet, this will call :meth:`~scvi.hub.HubModel.read_large_training_adata`,
        which will attempt to download from the source url. Otherwise, it will simply return the loaded data.
```